### ISSUE TITLE
Virtualbox shared folders broken, works in Ubuntu but not in Boot2docker

### ISSUE THREAD

**Issue by rdsubhas** [boot2docker/boot2docker#1069]
Virtualbox shared folders broken, checked with different images and machines

**NOTE:** This is not related to the age-old VirtualBox SendFile bug or #887. **Functionality works fine in Ubuntu but does not work in boot2docker.** Sorry but we can confirm this problem is isolated to boot2docker.

### boot2docker 1.8.2
* Start boot2docker 1.8.2 VM
* Go to settings and share a folder (pick any folder of choice)
* Mount the folder using vboxsf: `mkdir ~/share && sudo mount -t vboxsf HostFolderName ~/share`
* Create a file in host OS: `echo 1 > test.json`
  * Read the file in boot2docker: `cat ~/share/test.json`
  * Prints 1, fine for the first time.
* Update the file in host OS: `echo 2 > test.json`
  * Read the file in boot2docker: `cat ~/share/test.json`
  * **Still prints old "1"**
  * **It reads file correctly first time, but from then on the file becomes cached and updates don't reflect**

### ubuntu 14.04
* Start an Ubuntu 14.04 VM and do the same steps above
* **File updates are correctly picked up again and again, no issues**

**This does not seem to be a Virtualbox error, because things are working fine in an Ubuntu guest.** Only failing in boot2docker guest VM. Tried in two different machines both OSX Yosemite. Able to reproduce with latest boot2docker 1.8.2 and VirtualBox 5.0.0/5.0.4/5.0.6.

**Comment by rdsubhas** [comment]
I have screenshots of the Terminal in both cases, unable to upload since there seems to be an issue with github-cloud attachments, will upload very soon.

**Comment by rdsubhas** [comment]
Ubuntu - File contents are up to date:
![ubuntu-cloudimg-trusty-vagrant-amd64__running__and_1__zsh](https://cloud.githubusercontent.com/assets/1570745/10344735/9aef5430-6d42-11e5-9586-a53288085186.png)

Boot2docker - File contents are cached after a `cat` and then never updated:
![boot2docker__ssh](https://cloud.githubusercontent.com/assets/1570745/10344743/aaccb3d4-6d42-11e5-810e-518171620e13.png)

**Comment by rdsubhas** [comment]
Further investigation: Its got to do with the Filesystem cache/buffers. `cat` shows the cached file, wherease `less` and `vi` show the uncached file.

Running the following command clears the filesystem cache and shows the updated file:
```
echo 3 > /proc/sys/vm/drop_caches
```

I'm investigating with older versions of boot2docker ISO to find out whether this is a regression.

**Comment by trigun539** [comment]
Does anybody know a workaround for this? Have to perform that command every time I change a file.

**Comment by rdsubhas** [comment]
@trigun539 I'm testing this with `boot2docker 1.6.2` to see if the same shared folder problem is present there as well. could be a regression somewhere...

**Comment by rdsubhas** [comment]
**UPDATE 1:** Shared folder works perfectly with Boot2docker 1.6.2 release. No caching issues, file changes immediately  reflected.

<img width="637" alt="screen_shot_2015-10-12_at_11_13_36_am" src="https://cloud.githubusercontent.com/assets/1570745/10422060/a576ea88-70d2-11e5-83ae-06044ad10844.png">

<img width="1280" alt="screen_shot_2015-10-12_at_11_13_56_am" src="https://cloud.githubusercontent.com/assets/1570745/10422067/acb23f3c-70d2-11e5-85e2-349dc2c7b25b.png">

**Comment by rdsubhas** [comment]
**UPDATE 2: Gotcha:** Its *almost confirmed* to be the [FSCACHE changes](https://github.com/boot2docker/boot2docker/releases/tag/v1.7.0) in boot2docker 1.7.0.

<img width="634" alt="screen_shot_2015-10-12_at_11_20_44_am" src="https://cloud.githubusercontent.com/assets/1570745/10422136/a4af29de-70d3-11e5-8bd8-68c3c7711927.png">

<img width="1279" alt="screen_shot_2015-10-12_at_11_21_22_am" src="https://cloud.githubusercontent.com/assets/1570745/10422138/abfc3538-70d3-11e5-8a77-f0b81a19065c.png">

**Comment by rdsubhas** [comment]
pinging @tianon @paolomainardi @SvenDowideit to have a second look, referencing #699

I'm now disabling each `CONFIG_FSCACHE` related flag introduced in #699 and rebuilding the ISO to test

**Comment by braderhart** [comment]
I'm encountering a similar bug, although it could be related to the sendfile #887 bug referenced above. The problem is that even if I disable sendfile in my NGINX config that if I modify a file from the host that it still serves old content.

If I exec a shell in my container running NGINX and modify a file changes don't show either.

However, if I modify the file using a different container then the changes work. Before disabling sendfile, even modifying a file in a container would not show the changes.

What is the status of rebuilding the ISO and disabling `CONFIG_FSCACHE`? 

My suggestion is to use NFS instead of vboxsf in Boot2Docker. I am also willing to offer a bounty of $200 to whoever can get this fixed. Please let me know if you're interested and I'll post the bounty.

**Comment by rdsubhas** [comment]
@braderhart sorry was not able to do that yet, will be checking with `CONFIG_FSCACHE` today. Good to know that I'm not the only person facing this issue ;)

**Comment by braderhart** [comment]
Thank you for helping out. You're definitely not alone in this issue as I've found multiple reports referencing the same issues.

**Comment by braderhart** [comment]
Similar issue using an Alpine Linux image which is what I'm using for my NGINX container (specifically jekyll/jekyll:pages):

https://github.com/gliderlabs/docker-alpine/issues/84

**Comment by princemaple** [comment]
:+1:

**Comment by tianon** [comment]
I've pushed 200d65bc792ec89b9c48066d6e1c5a9d3642e4c1, and will be building `-rc2` including it later today, which will _hopefully_ fix this one. :+1:

**Comment by rdsubhas** [comment]
awesome @tianon will try out the RC :+1:

**Comment by tianon** [comment]
https://github.com/tianon/boot2docker-legacy/releases/tag/v1.9.0-rc2 :smile:

**Comment by rdsubhas** [comment]
checking it out right away

**Comment by tianon** [comment]
:heart:

**Comment by lalyos** [comment]
My usecase was running mkdocs in a container, and this issue was driving me crazy! I confirm its fixes the issue. Thanks @tianon

**Comment by rdsubhas** [comment]
@lalyos just to confirm: does `1.9.0-rc2` fix the issue? or running `cache-clear`?

**Comment by lalyos** [comment]
Yes, this what i did:
```
boot2docker  upgrade --iso-url=https://github.com/tianon/boot2docker-legacy/releases/download/v1.9.0-rc2/boot2docker.iso
```
aaaaand magically it works again. I had the same symptom you describe: `cat` was cashing while anything else like `wc` or `more` not.

**Comment by rdsubhas** [comment]
Good to hear! I still see `cat` is caching, and its nice to note that `mkdocs`, `less`, `vi`, `npm` don't use the cache. I'm checking with `python` in a container. Thanks a bunch @tianon but I'm not sure what to make of the output :) Any ideas?

<img width="1279" alt="screenshot_24_10_15_6_33_pm" src="https://cloud.githubusercontent.com/assets/1570745/10710606/9f795e1c-7a7e-11e5-8dac-9bdca394a668.png">

**Comment by tianon** [comment]
Interesting -- had you run "cache-clear" before that?  I wonder if
something inadvertently caused the "fscache" module to get loaded in the
kernel ("lsmod" should tell you).

**Comment by rdsubhas** [comment]
Yep ran the same steps multiple times just to make sure, and ran `cache-clear` in between. Also `lsmod` does not list "fscache" both in the RC as well as 1.8.2, guess fscache is built into the kernel when compiled.

<img width="1279" alt="shell" src="https://cloud.githubusercontent.com/assets/1570745/10710628/9dda1fdc-7a7f-11e5-928a-a51c04eb4c37.png">

**Comment by paolomainardi** [comment]
@tianon thanks for fixing it, seems that some refactoring on the PR made the fscache statically compiled instead of being a module, as it was supposed to be, sorry for that.

**Comment by tianon** [comment]
No worries, glad we got it :+1:

**Comment by braderhart** [comment]
I think something is broke in Docker Toolbox 1.9.0-rc2 for Windows with the new boot2docker ISO. I'm getting an error saying that virtualization extensions aren't enabled when I've verified that they are.

**Comment by TomBZombie** [comment]
I'm trying to solve this exact problem (Not using boot2docker but it seems to be a larger issue) I'm getting this in any Linux OS when:

1) Try to open a file which doesn't exist yet: `stat non-existant.file`
2) Create the file on the host
3) Now try to open the file on the guest: `cat non-existant.file` it says "no such file or directory" and the permissions/date all come up as ?????

If you write the file on the host before reading it on the guest it works correctly. Similarly, clearing the cache using drop_caches makes the files accessible, but this has to be run each time a file is affected. 

I've recompiled an Arch Kernel with various options for `CONFIG_FSCACHE` and the problem remains. I'd love to know what you did if you have managed to solve the problem.

I've posted a bug on the VirtualBox bug tracker here: https://www.virtualbox.org/ticket/14746 please update it if you have any more information that could help fix it.

**Comment by liviubalan** [comment]
If you want to learn more about shared folders on VirtualBox (Linux Mint/Ubuntu Desktop) you can checkout <a href="http://www.liviubalan.com/virtualbox-shared-folders-on-linux-mint">VirtualBox Shared Folders on Linux Mint (Video Tutorial)</a>.

**Comment by garykrige** [comment]
This is still an issue for me.

**Comment by rdsubhas** [comment]
@garykrige just a try, but have you checked with Virtualbox Guest Additions installed?

**Comment by zoidbergwill** [comment]
@rdsubhas Awesome, that seems to have fixed it.

**Comment by garykrige** [comment]
nah it didnt fix it for me. using uwsgi to reload python files, just not working.

**Comment by sylus** [comment]
Really happy to have found this issue, and exactly as above if I run the following command, then the files work but I would have to run the command after every change.

```sh
echo 3 > /proc/sys/vm/drop_caches
```

**Comment by sylus** [comment]
This issue is more then 6 months old and it does seem like there is a serious regression in Boot2Docker.

Is this something still being actively pursued or should I go straight to docker for mac / docker for windows to not need boot2docker.

**Comment by TomBZombie** [comment]
I worked out the problem, the kernel version used has this problem. Upgrading the kernel solves it.

**Comment by sylus** [comment]
So just to clarify I guess boot2docker would have to update its Linux Kernel version? I know you were on Arch Linux, but what kernel version worked for you?

**Comment by TomBZombie** [comment]
I believe it was 3.18 that had the problem. Upgrading fixed it, however that's from memory. That said, it was a year ago so any kernel less than a year old will have the problem resolved. Earlier kernels also worked for me, it was a small window of 2-3 kernel versions.

And to clarify it's the guest's kernel not the host's.

**Comment by sylus** [comment]
Unfortunately I am on boot2docker v1.12.3 which has a linux kernel of v4.4.27 so seems to still be an issue related to boot2docker. At least for now I can run `echo 3 > /proc/sys/vm/drop_caches` to sync everything. I'd happily look further into this if could be pointed to the right direction :)

**Comment by sylus** [comment]
Hopefully it is okay to add but as this seems to be a major issue pretty much resulting in not being able to successfully use docker + boot2docker in any type of dev workflow with volume mounts. As such shouldn't this issue have a higher priority label? I might even recommend that this issue is documented in the readme as I have spent hours assuming it was something I did wrong with the volume mounts.

Basically I was hoping we could allow the broader docker community to be aware of this issue with the hopes we could isolate the direction of the problem :)

**Comment by mcatalin1** [comment]
This issue is still present in Boot2Docker version 1.12.3

System:
-------------------------------------
boot2docker:
	Boot2Docker version 1.12.3, build HEAD : 7fc7575 - Thu Oct 27 17:23:17 UTC 2016
	Docker version 1.12.3, build 6b644ec

	Linux default 4.4.27-boot2docker #1 SMP Tue Oct 25 19:51:49 UTC 2016 x86_64 GNU/Linux

VirtualBox:
	Version 5.0.20 r106931

Host System:
	OS X El Capitan - 10.11.6

Affected container: 
	Alpine:3.4


How to replicate the issue
-------------------------------------

1. run docker quick start terminal
2. execute: 
	docker run -dt --name testfs -v $HOME:/data alpine:latest sh
	docker exec -ti testfs sh
	cd /data

	this will start a new Alpine container and mount your home directory as /data
	
3. on your *host* system create a text file in the home dir:
	cd
	echo "line1" > test_file.txt
	
4. in the *container* terminal:
	cat test_file.txt

	Result:
	line1	
	
5. on your *host* system create a text file in the home dir:
	echo "line2" >> test_file.txt

6. in the *container* terminal:
	cat test_file.txt

	Result:
	line1
	
	the result should be:
	line1
	line2	
	
	as we just added another line to the file.

**Comment by ehernandez-xk** [comment]
I have the same problem using 
- Boot2docker  1.12.6
- macOS Sierra, 10.12.4

I have spent hours trying to find what happen, I don't know if this doesn't happen in newest versions. I will try

**Comment by mcatalin1** [comment]
so the problem is not boot2docker but vboxsf (virtualbox shared folders) filesystem. 
to avoid this problem I'm mounting the home directory with NFS on mac and Samba (SMB) on windows. 

for mac use https://github.com/adlogix/docker-machine-nfs
you'll get faster speeds and reliability

for Windows
1. share the folder on the host 
2. mount it with something like this:
mount -t cifs -o username=USERNAME,password=PASSWD //192.168.1.88/shares /mnt/share

more info here:  https://askubuntu.com/questions/101029/how-do-i-mount-a-cifs-share

to make this permanent add it to /var/lib/boot2docker/bootlocal.sh

**Comment by ehernandez-xk** [comment]
Tks @mcatalin1 

You are right. NFS works well, I decided to use vboxsf because works on Mac and Windows, and NFS requires Administrator privileges every time you create a mount. (there are cases when the user does not have sudo power) for that reason I'm looking a way to prevent this issue. 😸

**Comment by oraoto** [comment]
It seems fixed in boot2docker v17.06.0-ce.

Environment:

+ boot2docker v17.06.0-ce
+ Windows 10 1703
+ VirtualBox 5.1.22

![image](https://user-images.githubusercontent.com/24709398/27766857-000fa95a-5f13-11e7-9dfc-03edf9035040.png)

**Comment by ehernandez-xk** [comment]
Yea, you are right, I tested it with:
 
- Vagrant 1.9.6
- VB 5.1.22r115126

Thank you!

**Comment by deAtog** [comment]
This is a bug in the Virtual Box Guest Additions for Linux. A patch has recently been accepted to correct this issue for Linux kernels > 2.6.30. For more information see bugs: [#819](https://www.virtualbox.org/ticket/819), [#17053](https://www.virtualbox.org/ticket/17053), [#9069](https://www.virtualbox.org/ticket/9069), [#12597](https://www.virtualbox.org/ticket/12597), and [#14920](https://www.virtualbox.org/ticket/14920). With any luck the fix for this issue should be in the Virtual Box 5.2.1 release. Relevant change set is here: [#69165](https://www.virtualbox.org/changeset/69165/vbox)

**Comment by deAtog** [comment]
@tianon This issue is fixed in VirtualBox Guest Additions 5.2.2. Please update accordingly. Thanks.
